### ISSUE TITLE
Project dependencies may have API risk issues

### ISSUE THREAD

**Issue by PyDeps** [apple/ml-hypersim#50]
Hi, In **ml-hypersim**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using

```
h5py=2.10.0=py37h3134771_0
joblib=0.17.0=py_0
matplotlib=3.3.1=0
pandas=1.1.3=py37hb1e8313_0
python=3.7.9=h26836e1_0
scikit-learn=0.23.2=py37h959d312_0
scipy=1.5.2=py37h912ce22_0
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **h5py** can be changed to *>=2.5.0,<=3.7.0*.
The version constraint of dependency **joblib** can be changed to *==0.7.0d*.
The version constraint of dependency **joblib** can be changed to *>=0.3.6.dev,<=1.1.0*.
The version constraint of dependency **pandas** can be changed to *==0.6.0*.
The version constraint of dependency **pandas** can be changed to *>=0.13.0,<=0.23.4*.
The version constraint of dependency **scikit-learn** can be changed to *>=0.9,<=0.20.4*.
The version constraint of dependency **scipy** can be changed to *>=0.12.0,<=0.15.1*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

<details><summary>The calling methods from the h5py</summary>
 <pre>h5py.File
</pre>
</details>
<details><summary>The calling methods from the joblib</summary>
 <pre>joblib.Parallel
joblib.delayed
</pre>
</details>
<details><summary>The calling methods from the pandas</summary>
 <pre>pandas.DataFrame.to_csv
pandas.DataFrame.from_dict
pandas.read_csv
</pre>
</details>
<details><summary>The calling methods from the scikit-learn</summary>
 <pre>sklearn.preprocessing.normalize
joblib.Parallel
joblib.delayed
</pre>
</details>
<details><summary>The calling methods from the scipy</summary>
 <pre>linalg.norm
scipy.linalg.block_diag
linalg.det
</pre>
</details>
<details><summary>The calling methods from the all methods</summary>
 <pre>os.path.realpath
sqrt.astype
array.append
loaded_prefix_str.len.loaded_line.strip
face_vni.append
renderer.classes.SettingsEnvironment.getInstanceOrCreate
depth_meters.astype
xscale
df.append.to_csv
ax.set_title
self._material_names.append
renderer.classes.SettingsVFB.getInstances
N_bump_cam.T.A.reshape.astype
pandas.read_csv.isin
vray_render_entity_id.astype.astype
tmp_path_prefixes.append
mayavi_utils.points3d_color_by_scalar
renderer.classes.BRDFVRayMtl
pandas.read_csv
inspect.getfile
dstack.astype
zipfile.ZipFile
in_scene_fileroot.detail_dir_when_rendering.os_path_module_when_rendering.join.replace
octomap_utils.generate_octomap_samples
renderer.classes.BitmapBuffer.getInstances
mayavi.mlab.axes
renderer.classes.UVWGenChannel
color_hist_norm_bin_centers_X.ravel
ones
hist
np.random.rand
mesh_vertices_sii_.shape.arange.reshape
os.makedirs
WebFile
df_node_strings.append.append
vray.VRayRenderer.export
t.strip
sqrt
sklearn.preprocessing.normalize
df_nodes.append.append
nodes_to_visit.pop.__getitem__
in_line.replace.replace.replace.split
metadata_camera_trajectories_csv_file.pd.read_csv.rename_axis.reset_index
ax.set_xticks
scipy.spatial.cKDTree
N_cam.T.A.reshape.astype
mayavi.mlab.show
renderer.classes.SettingsGI.getInstances
savefig
self._material_files.append
renderer.classes.SettingsDMCSampler.getInstances
np.random.choice
args.in_mesh_metadata_objects_file.pd.read_csv.rename_axis.reset_index.set_index
normal_cam_.reshape
pandas.read_csv.to_records
asarray
in_filename.split
cross
renderer.classes.CameraPhysical.getInstances
obj_data.material_names.df_materials.loc.rename_axis.reset_index
xticks
renderer.classes.SettingsCamera.getInstanceOrCreate
obj_file_utils.load_obj_file
self._material_names.pd.Series.drop_duplicates
np.power
approx_mvbb_utils.generate_oriented_bounding_box_2d
dfi.Index.df_channels.loc.copy
pixel_y.pixel_x.dstack.reshape
zipfile.ZipFile.infolist
vray.List
line.strip.split
histogram
renderer.classes.SettingsOutput.getInstanceOrCreate
face_vi.append
cv2.imread.copy
zip
ylim
transform_point_world_from_obj
renderer.classes.SettingsDMCSampler.getInstanceOrCreate
show
face_vti.append
array
subplot
H.astype
print
self.material_param_map.update
renderer.classes.SettingsImageSampler.getInstances
renderer.classes.RenderChannelNodeID
os.system
renderer.classes.GeomStaticMesh
float
linalg.norm
plt.figure.tight_layout
normal_hist.astype
log
datetime.datetime.strptime
renderer.classes.SettingsIrradianceMap.getInstances
join
np.percentile
range
renderer.classes.LightDome.getInstances
args.in_mesh_metadata_objects_file.pd.read_csv.rename_axis
renderer.classes.TexVectorToColor
files_to_visit.append
self._vertices.append
self._normals.append
round
tmp_nodes_to_visit_added_by.append
normal_cam.copy.reshape
scene_name.startswith
os.path.isfile
self.available
renderer.classes.SettingsLightCache.getInstanceOrCreate
p_test_screen.astype
process_scene
xlabel
rgb_to_brightness
num_pts.ones.colors.c_.astype
np.sum
ylabel
in_line.strip
arctan2
start_camera_positions.pd.DataFrame.drop_duplicates
yticks
nodes_to_visit_added_by.pop
pandas.read_csv.to_csv
argmin
argparse.ArgumentParser.parse_args
normal_world_.reshape
normal_cam_1d_.reshape.reshape
loaded_time.unloaded_time.total_seconds
self._faces.append
renderer.classes.SettingsUnitsInfo.getInstances
where
glob.glob
obj_data.material_names.df_materials.loc.rename_axis
renderer.classes.SettingsCameraDof.getInstanceOrCreate
any
distutils.dir_util.copy_tree
path_utils.add_path_to_sys_path
renderer.classes.SettingsEXR.getInstanceOrCreate
mayavi_utils.points3d_color_by_rgb_value
cv2.imread
renderer.classes.TexCombineColor
df_cam_.to_records
df_node_strings.append.to_csv
np.random.randint
compute_ray_camera_space
in_line.replace
mesh_faces_vi_sii.ravel
path_prefixes.pop
mtl_file
np.random.shuffle
in_camera_name.lstrip
n_jobs.Parallel
identity
color_hist_denorm_bin_centers_X.ravel
np.maximum
arange
renderer.classes.TexSampler.getName
unravel_index
renderer.classes.RenderChannelDenoiser.getInstanceOrCreate
in1d
nodes_to_visit.pop.getMeta
ax.set_yticks
vray.VRayRenderer.setOnLogMessage
generate_fragment
df.reset_index.to_records
get_strings_for_plugin
renderer.classes.RenderChannelExtraTex
len
sphere_sampling_utils.generate_evenly_distributed_samples_on_sphere
renderer.classes.Node
vray.VRayRenderer
logical_and
mesh_vertices_si_.shape.arange.reshape
entry.is_dir
histogramdd
df.df.to_numpy
mayavi.mlab.xlabel
_vray_user_params._set_vray_user_params
os.path.splitext
self._object_names.pd.Series.drop_duplicates.tolist
transform_point_screen_from_obj
np.sqrt
matplotlib.colors.hsv_to_rgb
renderer.classes.SettingsColorMapping.getInstances
df.df.loc.to_numpy
xlim
df_cameras.append.to_csv
n.str.split
metadata_camera_trajectories_csv_file.pd.read_csv.rename_axis
os.path.abspath
dot
pandas.DataFrame.to_csv
N_bump_cam.T.A.reshape
in_line.strip.startswith
os.chdir
barh
os.getcwd
generate_plugins_for_quad
f.replace
mayavi.mlab.triangular_mesh
metadata_camera_trajectories_csv_file.pd.read_csv.rename_axis.to_records
metadata_rendering_tasks_csv_file.pd.read_csv.rename_axis
fnmatch.fnmatch
renderer.classes.MtlSingleBRDF
np.arctan
np.minimum
argsort
np.max
vray.VRayRenderer.close
renderer.classes.SettingsLightCache.getInstances
obj_file
renderer.classes.SettingsVFB.getInstanceOrCreate
sort
unloaded_prefix_str.len.unloaded_line.strip
semantic_id_to_name_map.keys
self._material_names.pd.Series.drop_duplicates.tolist
os_path_module_when_rendering.join
ax.imshow
intersection_distances.reshape.reshape
set
diffuse_reflectance_hist.astype
sin
obj_data.material_names.df_materials.loc.rename_axis.reset_index.to_csv
tmp_nodes_to_visit.append
f.astype.reshape
ax.cax.colorbar
self._group_names.pd.Series.drop_duplicates
approx_mvbb_utils.generate_oriented_bounding_box_3d
np.min
cumsum
cos
normal_world_1d_.reshape.reshape
matplotlib.colors.rgb_to_hsv
sqrt.sum
PIL.Image.fromarray
in_line.replace.replace
renderer.classes.RenderChannelDenoiser.getInstances
newaxis.u_01.newaxis.v_01.newaxis.w_01.dstack.astype
log.sum
dfi.Index.df.loc.copy
start_camera_positions.pd.DataFrame.drop_duplicates.to_numpy
logspace
log.copy
mayavi.mlab.quiver3d
matrix
plt.figure
H.sum
diag
normal_cam_1d_.reshape.copy
renderer.classes.SettingsUnitsInfo.getInstanceOrCreate
traversal_stack.append
abs
random_walk_utils.generate_camera_trajectory_random_walk
camera_name_current.scene_name_current.frames_ids_excluded_flagged.append
f.astype
renderer.classes.CameraPhysical.getInstanceOrCreate
ax.spines.set_visible
ceil
mesh_faces_oi.oi.mesh_faces_vi.ravel
N_cam.T.A.reshape
shutil.copy
args.in_mesh_metadata_objects_file.pd.read_csv.rename_axis.reset_index
H.sum.ravel
mayavi.mlab.draw
int
gca.get_yticklabels
mayavi.mlab.zlabel
in_filename.replace
mesh_faces_vi_si.ravel
pandas.read_csv.itertuples
isfinite
rgb_to_hsv_xyz
normal_bump_world_.reshape
prod
f.matrix.A1.astype
pandas.Series
camera_name.in_scene_fileroot.detail_dir_when_rendering.os_path_module_when_rendering.join.replace
renderer.classes.LightRectangle.getInstances
colorbar
renderer.classes.FilterArea.getInstanceOrCreate
tmp_lines.append
diffuse_illumination_hist.astype
download
renderer.classes.TexVectorToColor.getName
cv2.imwrite
v.split
imsave
os.path.join
np.random.randn
set.append
df_camera_parameters.append.append
unique
logical_not
generate_fragments_for_line
allclose
min
vray_render_entity_id.astype
vray.VRayRenderer.load
ones_like
log_H.sum.ravel
out_file.write
color_hist_denorm_bin_centers_Z.ravel
mayavi.mlab.plot3d
mayavi.mlab.ylabel
np.random.seed
self._object_names.append
download_files
isinstance
generate_plugins_for_quad.append
gca
setp
argparse.ArgumentParser.add_argument
os.path.basename
dstack
requests.session.head
nodes_to_visit.append
renderer.classes.SettingsIrradianceMap.getInstanceOrCreate
exit
os.rmdir
line_.strip.startswith
renderer.classes.Node.getInstances
dict
args.out_light_cache_file.replace
joblib.delayed
pandas.DataFrame
gca.invert_yaxis
args.out_irradiance_map_file.replace
color_hist_denorm_bin_centers_Y.ravel
vray.Transform
np.linalg.norm
tuple
sys.path.insert
embree_utils.generate_ray_intersections
process_scene.delayed
files_to_visit.pop
clip
logical_and.reshape
f.readlines
rgb_color_hist.astype
PIL.ImageDraw.Draw
argmax
metadata_camera_trajectories_csv_file.pd.read_csv.rename_axis.reset_index.set_index
renderer.classes.TexSampler
t.isdigit
str
PIL.ImageDraw.Draw.line
renderer.classes.SettingsGI.getInstanceOrCreate
requests.session
in_line.replace.replace.replace
df_images_public.scene_name.unique
zipfile.ZipFile.extract
paths.append
metadata_camera_trajectories_csv_file.pd.read_csv.to_records
all
sphere_sampling_utils.generate_evenly_distributed_samples_on_sphere.astype
os.listdir
renderer.classes.SettingsColorMapping.getInstanceOrCreate
valid_mask.reshape
df_camera_parameters.append.to_csv
sum
scipy.spatial.cKDTree.query
sklearn.preprocessing.normalize.reshape
semantic_id_to_name_map.items
line.strip.strip
renderer.classes.MtlWrapper.getInstances
pandas.DataFrame.from_dict
vray.Color
renderer.classes.TexBitmap
self._group_names.pd.Series.drop_duplicates.tolist
vray.AColor
np.array
generate_plugins_for_quad.reverse
color_hist_norm_bin_centers_Y.ravel
shlex.split
traversal_stack.pop
logical_or
residual_hist.astype
message.startswith
self._texcoords.append
renderer.classes.LightMesh.getInstances
args.frames.split
max
line_.strip
joblib.Parallel
mayavi.mlab.outline
os.path.dirname
out_lines.append
df.append.append
renderer.classes.RenderChannelColor
strings.append
path_utils.get_current_source_file_path
os.path.isdir
tan
f.create_dataset
arctan
open
in_line.replace.split
color_hist_norm_bin_centers_Z.ravel
df.included_in_public_release.value_counts
inspect.currentframe
np.abs
transform_point_screen_from_world
ones_like.astype
residual.astype
vray.Vector
title
df.reset_index.reset_index
linspace
matplotlib.rcParams.update
node.getMeta.keys
scenes.append
imread
dict.keys
prim_ids.reshape.reshape
np.matrix
shutil.move
np.linalg.det
np.identity
plot
renderer.classes.SettingsOutput.getInstances
metadata_rendering_tasks_csv_file.pd.read_csv.rename_axis.reset_index.set_index
os.path.exists
renderer.classes.SettingsCamera.getInstances
colors.astype
line.strip.startswith
scipy.linalg.block_diag
mpl_toolkits.axes_grid1.ImageGrid
list
self._group_names.append
pandas.read_csv.drop_duplicates
self.session.get
normal_bump_cam_.reshape
isclose
time.sleep
octomap_utils.generate_octomap
normal_world.reshape.reshape
remove_plugin_names.append
valid_mask.copy
re.split
nodes_to_visit_added_by.append
path_prefixes.append
metadata_rendering_tasks_csv_file.pd.read_csv.rename_axis.reset_index
argparse.ArgumentParser
imshow
diff
renderer.classes.BRDFVRayMtl.getInstances
mayavi.mlab.figure
in_file.readlines
in_camera_name.lstrip.isdigit
main
logical_not.reshape
h5py.File
self._object_names.pd.Series.drop_duplicates
linalg.det
pandas.read_csv.to_numpy
os_path_module_when_rendering.join.replace
zeros
zeros_like
count_nonzero
np.tan
exp
linspace.isdigit
renderer.classes.RenderView.getInstanceOrCreate
os.system.check_returncode
renderer.classes.SettingsImageSampler.getInstanceOrCreate
vray.Matrix
renderer.classes.SettingsMotionBlur.getInstanceOrCreate
renderer.classes.BitmapBuffer
df_nodes.append.to_csv
camera_name.startswith
meshgrid
dstack.reshape
subprocess.run
camera_keyframe_frame_indices.astype
df_cameras.append.append
load_mtl_file
f.write
nodes_to_visit.pop
</pre>
</details>

@developer
Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.

**Comment by mikeroberts3000** [comment]
Hi @PyDeps, do you have a proposal for what our requirements file should look like?
